doc/coding.md

```diff
@@ -0,0 +1,333 @@
+see the distribution of languages, which can be achieved with
+`app.metrics.unique()`.
+
+```javascript
+router.get('/:name/news/:lang?', function(req, res) {
+
+    // set the language if not set
+    var lang = req.params.lang || 'en';
+
+    // count the language occurrence
+    app.metrics.unique('people.news.lang', lang);
+    // start measuring the time
+    var startTime = Date.now();
+
+    // get the news
+    return preq.get({
+        uri: 'https://' + lang + '.wikinews.org/wiki/'
+                + encodeURIComponent(req.params.name)
+    }).then(function(wnRes) {
+        // external request done, report the request time
+        app.metrics.timing('people.news.time', Date.now() - startTime);
+        // also report the payload's size
+        app.metrics.histogram('people.news.size', wnRes.body.length);
+        res.status(200).type('html').send(wnRes.body);
+    });
+
+});
+```
+For more information on the available methods, see the [service-runner
+documentation](https://github.com/wikimedia/service-runner#metric-reporting).
+
+## Test Cases
+
+The service needs to thoroughly tested since other services and clients are
+going to depend on it. The template uses
+[mocha](https://github.com/mochajs/mocha) for test execution and provides some
+useful utility functions in [test/utils](../test/utils).
+
+To create a test suite for our example routes, create the `people` directory in
+`/test/features/` and two files inside of it: `about.js` and `news.js`. These
+will test the example routes. Let's start with `about.js`:
+
+```javascript
+'use strict';
+
+
+// mocha defines to avoid JSHint breakage
+/* global describe, it, before, beforeEach, after, afterEach */
+
+
+var preq   = require('preq');
+var assert = require('../../utils/assert.js');
+var server = require('../../utils/server.js');
+
+
+describe('people - about', function() {
+
+    this.timeout(20000);
+
+    before(function () { return server.start(); });
+
+    // common URI prefix
+    var uri = server.config.uri + 'en.wikipedia.org/v1/people/';
+
+    it('get HTML for index', function() {
+        return preq.get({
+            uri: uri + 'index/about'
+        }).then(function(res) {
+            // check the status
+            assert.status(res, 200);
+            // check the returned Content-Type header
+            assert.contentType(res, 'text/html');
+            // inspect the body
+            assert.notDeepEqual(res.body, undefined, 'No body returned!');
+        });
+    });
+
+    it('fail for a non-existent person', function() {
+        return preq.get({
+            uri: uri + 'Walt_Disney/about'
+        }).then(function(res) {
+            // if we are here, no error was thrown, not good
+            throw new Error('Expected an error to be thrown, got status: ', res.status);
+        }, function(err) {
+            // inspect the status
+            assert.deepEqual(err.status, 404);
```

Comment:
Amazing, that was exactly what I was looking for in https://github.com/gwicke/preq/issues/1 I'll read the rest later